### ISSUE TITLE
parallel builds of the same job

### ISSUE THREAD

**Issue by asherungar-ceva** [null511/Jenkins.NET#5]
I am heavily testing your features today :) - by the way, great tool, seems exactly what i needed and had much trouble looking for something like this.
thanks for the quick reply.

now i have an issue of running builds in parallel.
my job is a pipeline one, and "Do not allow concurrent builds" is not set.

i try to run the same job in parallel (with different params) but it seems as though they are waiting for one another instead of running in parallel.
any ideas?
thanks
Asher

**Comment by null511** [comment]
> I am heavily testing your features today :)

Thank you, it is much appreciated.

> by the way, great tool, seems exactly what i needed and had much trouble looking for something like this.

I'm glad it has been useful. For my purposes, I wanted to leverage Jenkins and Octopus as a CI pipeline, but wrapped in my own web-interface for management purposes.

> i try to run the same job in parallel (with different params) but it seems as though they are waiting for one another instead of running in parallel.

This should work fine as long as you are using separate `JenkinsJobRunner` instances. Each instance will manage a unique queue and build number for the associated job. Starting a second job using the same runner would overwrite the ID of the first item. I will probably add additional error handling to prevent the user from starting more than one job from a runner.

**Comment by asherungar-ceva** [comment]
Ok, I think I see what the problem is.
I’m triggering the job from a TFS (XAML) build, where “parallel” isn’t really parallel.

I guess in this case you have the “Async” method? So I can trigger a job, exit the class, and return again with a new client instance to poll the jobs’ status?
Again, missing some example of how to do this in the documentation.

Any chance you can share some example on how to achieve it?

Thanks,

Asher Ungar
Debugger Team Leader
SDT,        CEVA-DSP

**Comment by null511** [comment]
I'm having trouble following your usage, could you provide a more detailed example? Also, the `JenkinsJobRunner` was designed as a convenience class for easy use. For more advanced scenarios, it may be more useful to leverage the `JenkinsClient` directly. You can view the source of the `JenkinsJobRunner` as an [example](https://github.com/null511/Jenkins.NET/blob/master/Jenkins.NET/JenkinsJobRunner.cs).

> Ok, I think I see what the problem is. I’m triggering the job from a TFS (XAML) build, where “parallel” isn’t really parallel.

Are you using **TFS** > **Jenkins.NET** > **Jenkins**? If so, are you using Jenkins.NET from a web service which has a single instance, or a desktop/console application which will have an instance per-running-job?

-----

To run multiple jobs from a single application using `JenkinsJobRunner`, you will need to create an instance of `JenkinsJobRunner` for each running Job. You only need one instance of `JenkinsClient`.

```c#
var client = new JenkinsClient {...};

var jobName = "Test Job";
var taskList = new Task[3];
for (var i = 0; i < 3; i++) {
    var runner = new JenkinsJobRunner(client) {
        ...
    };

    taskList[i] = runner.RunAsync(jobName);
}

Task.WaitAll(taskList);
```

The above example should start the same job three time using asynchronous tasks, and then wait for all the tasks to complete.

**Comment by asherungar-ceva** [comment]
I wrapped the Jenkins.NET in a TFS XAML build custom activity (also C#).
The problem I had is that I need to exit the activity in order to parallel builds (because they are not really parallel in TFS XAML workflow).

So, I took your suggestion and I’m now using the client directly to queue a build, get it’s “queue Item number” and then exit the activity, and enter a new activity which checks the status of the build according to the item number.
Seems to work well ☺

Thanks

Asher Ungar
Debugger Team Leader
SDT,        CEVA-DSP

**Comment by asherungar-ceva** [comment]
I had to get the build number instead of using the “queuedItemNumber” – due to Jenkins issue:
https://issues.jenkins-ci.org/browse/JENKINS-31039

it turns out that after 5 minutes, Jenkins API deletes the queued item number information and results in “Not Found - 404” error.

Asher Ungar
Debugger Team Leader
SDT,        CEVA-DSP

From: Asher Ungar
Sent: Monday, February 26, 2018 09:36
To: 'null511/Jenkins.NET' <reply@reply.github.com>; null511/Jenkins.NET <Jenkins.NET@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [null511/Jenkins.NET] parallel builds of the same job (#5)

I wrapped the Jenkins.NET in a TFS XAML build custom activity (also C#).
The problem I had is that I need to exit the activity in order to parallel builds (because they are not really parallel in TFS XAML workflow).

So, I took your suggestion and I’m now using the client directly to queue a build, get it’s “queue Item number” and then exit the activity, and enter a new activity which checks the status of the build according to the item number.
Seems to work well ☺

Thanks

Asher Ungar
Debugger Team Leader
SDT,        CEVA-DSP
